R/abm.R

```diff
@@ -327,6 +327,102 @@ abm_table5 <- function(con, unit_code, analysis_start = abm_config()$start_year,
+}
+
+
+#' Retrieve Table 6 (OA data) for ABM
+#' 
+#' @param con connection to db, default is to use mssql connection
+#' @param unit_code for filtering on one or more unit code(s), which can be KTH, a one letter school code, an integer department code or a KTH-id (optional)
+#' @return tibble with OA-status of all publications from selected organizational units
+#' @import DBI dplyr tidyr purrr
+#' @export
+abm_table6 <- function(con = con_bib(), unit_code) {
+    
+    oa_data <- abm_oa_data(con =con, unit_code = unit_code)
+    
+    # Year-dependent part of table
+    table1 <-
+        oa_data %>%
+        group_by(Publication_Year) %>%
+        filter(is_oa=="TRUE" | is_oa=="FALSE") %>%
+        collect() %>%
+        summarise(P_tot=n(),
+                  oa_count=sum(as.logical(is_oa), na.rm=TRUE),
+                  gold_count=sum(as.logical(oa_status=="gold"), na.rm=TRUE),
+                  hybrid_count=sum(as.logical(oa_status=="hybrid"), na.rm=TRUE),
+                  green_count=sum(as.logical(oa_status=="green"), na.rm=TRUE),
+                  bronze_count=sum(as.logical(oa_status=="bronze"), na.rm=TRUE),
+                  closed_count=sum(as.logical(oa_status=="closed"), na.rm=TRUE),
+                  oa_share=mean(as.logical(is_oa), na.rm=TRUE)) %>%
+        ungroup() %>%
+        mutate(Publication_Year_ch = as.character(Publication_Year)) %>%
+        arrange(Publication_Year_ch) %>%
+        select(Publication_Year_ch, P_tot, oa_count, gold_count, hybrid_count, green_count, bronze_count, closed_count, oa_share)
+    
+    # Insert blank years
+    years_to_add <- table1[!(as.character(as.numeric(table1$Publication_Year_ch)+1) %in% table1$Publication_Year_ch)
+                           & !is.na(lead(table1$Publication_Year_ch)),"Publication_Year_ch"] %>%
+        mapply(function (x) as.numeric(x)+1, .)
+    
+    for (year in years_to_add){
+        if (length(year) > 0){
+            plusyear = 1
+            while (!(as.character(year+plusyear) %in% table1$Publication_Year_ch)){
+                years_to_add <- append(years_to_add, year+plusyear)
+                plusyear <- plusyear + 1
+            }
+        }
+    }
+
+    for (year in years_to_add){
+        if (length(year) > 0){
+            table1 <- table1 %>% rbind(data.frame(Publication_Year_ch = as.character(year),
+                                                  P_tot = integer(1),
+                                                  oa_count = integer(1),
+                                                  gold_count = integer(1),
+                                                  hybrid_count = integer(1),
+                                                  green_count = integer(1),
+                                                  bronze_count = integer(1),
+                                                  closed_count = integer(1),
+                                                  oa_share = 0))
+        }
+    } 
```

Comment:
For the "blank years"-step, this could potentially be refactored to use `purrr` iterators, magrittr-style chained "pipe" operators for more tidyverse style transformations. Obviously that is not necessary since base fcns can do what's needed, so more of a stylistic thing.

Reply:
I guess easiest way to add missing years would be something like `data.frame(year = abm_config()$start_year:abm_config()$stop_year)` and then left join this with the OA table. Do we really want oa_share to be = 0 for missing years?